// @wppconnect-team/wppconnect/dist/api/layers/labels.layer.js
"use strict";
/*
 * This file is part of WPPConnect.
 *
 * WPPConnect is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * WPPConnect is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with WPPConnect.  If not, see <https://www.gnu.org/licenses/>.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LabelsLayer = void 0;
const helpers_1 = require("../helpers");
const lists_layer_1 = require("./lists.layer");
class LabelsLayer extends lists_layer_1.ListsLayer {
    page;
    constructor(page, session, options) {
        super(page, session, options);
        this.page = page;
    }
    /**
     * Create New Label
     * @category Labels
     *
     * @example
     * ```javascript
     * client.addNewLabel(`Name of label`);
     * //or
     * client.addNewLabel(`Name of label`, { labelColor: '#dfaef0' });
     * //or
     * client.addNewLabel(`Name of label`, { labelColor: 4292849392 });
     * ```
     * @param name Name of label
     * @param options options of label
     */
    async addNewLabel(name, options) {
        return await (0, helpers_1.evaluateAndReturn)(this.page, ({ name, options }) => {
            WPP.labels.addNewLabel(name, options);
        }, { name, options });
    }
    /**
     * Add or delete label of chatId
     * @category Labels
     *
     * @example
     * ```javascript
     * client.addOrRemoveLabels(['[number]@c.us','[number]@c.us'],
     * [
     *   { labelId:'76', type:'add' },
     *   { labelId:'75', type:'remove' }
     * ]);
     * //or
     * ```
     * @param chatIds ChatIds
     * @param options options to remove or add
     */
    async addOrRemoveLabels(chatIds, options) {
        return await (0, helpers_1.evaluateAndReturn)(this.page, ({ chatIds, options }) => {
            WPP.labels.addOrRemoveLabels(chatIds, options);
        }, { chatIds, options });
    }
    /**
     * Get all Labels
     *
     * @example
     * ```javascript
     * client.getAllLabels();
     * ```
     */
    async getAllLabels() {
        return (0, helpers_1.evaluateAndReturn)(this.page, () => WPP.labels.getAllLabels());
    }
    /**
     * Get Label by id
     * @category Labels
     * @param id - Id of label
     *
     * @example
     * ```javascript
     * client.getLabelById('1');
     * ```
     */
    async getLabelById(id) {
        return await (0, helpers_1.evaluateAndReturn)(this.page, ({ id }) => {
            WPP.labels.getLabelById(id);
        }, { id });
    }
    /**
     * Delete all Labels
     * @category Labels
     *
     * @example
     * ```javascript
     * client.deleteAllLabels();
     * ```
     */
    async deleteAllLabels() {
        return await (0, helpers_1.evaluateAndReturn)(this.page, () => {
            WPP.labels.deleteAllLabels();
        });
    }
    /**
     * Add or delete label of chatId
     * @category Labels
     *
     * @example
     * ```javascript
     * client.deleteLabel();
     * ```
     * @param id Id or string to labels to delete
     */
    async deleteLabel(id) {
        return await (0, helpers_1.evaluateAndReturn)(this.page, ({ id }) => {
            WPP.labels.deleteLabel(id);
        }, { id });
    }
}
exports.LabelsLayer = LabelsLayer;
